api/response.go

```diff
@@ -0,0 +1,84 @@
+package api
+
+import (
+	"bytes"
+	"errors"
+	"github.com/gin-gonic/gin"
+	"log"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"strings"
+)
+
+type ErrorInfo struct {
+	Error string `json:"error"`
+}
+
+func writeResponse(ctx *gin.Context, metadata []byte, data []byte) {
+	response := &bytes.Buffer{}
+	writer := multipart.NewWriter(response)
+
+	err := writeData(ctx, writer, "application/json", metadata)
+	if err != nil {
+		ctx.AbortWithError(http.StatusInternalServerError, err)
+		return
+	}
+	err = writeData(ctx, writer, "application/octet-stream", data)
+	if err != nil {
+		ctx.AbortWithError(http.StatusInternalServerError, err)
+		return
+	}
+	err = writer.Close()
+	if err != nil {
+		log.Println(err)
+		ctx.AbortWithError(http.StatusInternalServerError,
+			errors.New("unexpected internal error when writing Response Data (close). Contact SysAdmin"))
+		return
+	}
+	ctx.Data(http.StatusOK, "multipart/mixed; boundary="+writer.Boundary(), response.Bytes())
+}
+
+func writeData(ctx *gin.Context, writer *multipart.Writer, contentType string, data []byte) error {
+	dataPart, err := writer.CreatePart(textproto.MIMEHeader{"Content-Type": {contentType}})
+	if err != nil {
+		log.Println(err)
+		return errors.New("unexpected internal error when writing Response Data (create part). Contact SysAdmin")
+	}
+	_, err = dataPart.Write(data)
+	if err != nil {
+		log.Println(err)
+		return errors.New("unexpected internal error when writing Response Data (write part). Contact SysAdmin")
+	}
+	return nil
+}
+
+func ErrorHandler(ctx *gin.Context) {
+	ctx.Next()
+
+	hasError := len(ctx.Errors) != 0 || ctx.Writer.Status() != http.StatusOK
```

Comment:
For example without that accessing /cats will return 404 without error message field. And I believe that during the meeting we agreed that we want every single error to return this structure, even if error itself will be empty (as it is in case of 404 because I didn't want to add any useless general messages).

So now actually the empty error list is kinda redundant condition, but I would like to leave both because we want to declare error if any of these conditions stands, regardless of the likelihood of one appearing without the other.

Reply:
Ok, thanks! Maybe add a comment about the writer status? It's not obvious at all that it's triggered by gin 404's (at least not to me 😄 )